src/init.c

```diff
@@ -451,52 +452,81 @@ static char *abspath(const char *in)
 }
 
 static void jl_resolve_sysimg_location(JL_IMAGE_SEARCH rel)
-{   // this function resolves the paths in jl_options to absolute file locations as needed
-    // and it replaces the pointers to `julia_home`, `julia_bin`, `image_file`, and output file paths
-    // it may fail, print an error, and exit(1) if any of these paths are longer than PATH_MAX
+{   // This function resolves the paths in jl_options to absolute file locations as needed
+    // and it replaces the pointers to `julia_home`, `julia_bin`, `julia_lib`, `image_file`,
+    // and output file paths.
+    // It may fail, print an error, and exit(1) if any of these paths are longer than PATH_MAX.
     //
     // note: if you care about lost memory, you should call the appropriate `free()` function
     // on the original pointer for each `char*` you've inserted into `jl_options`, after
     // calling `julia_init()`
     char *free_path = (char*)malloc(PATH_MAX);
     size_t path_size = PATH_MAX;
+
+    // set julia_bin
     if (uv_exepath(free_path, &path_size)) {
         jl_error("fatal error: unexpected error while retrieving exepath");
     }
     if (path_size >= PATH_MAX) {
         jl_error("fatal error: jl_options.julia_bin path too long");
     }
-    jl_options.julia_bin = (char*)malloc(path_size+1);
-    memcpy((char*)jl_options.julia_bin, free_path, path_size);
-    ((char*)jl_options.julia_bin)[path_size] = '\0';
+    jl_options.julia_bin = abspath(free_path);
+
+    // set julia_lib
+#ifdef _OS_WINDOWS_
+    HMODULE hModule = NULL;
+    if (!GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS,
+                           (LPCTSTR)jl_resolve_sysimg_location,
+                           &hModule)) {
+        jl_error("fatal error: unexpected error while retrieving library handle");
+    }
+    path_size = GetModuleFileNameA(hModule, free_path, PATH_MAX);
+    if(GetLastError() != ERROR_SUCCESS) {
+        jl_error("fatal error: unexpected error while retrieving libpath");
+    }
+#else // POSIX
+    Dl_info ifo;
+    if(!dladdr((void*)jl_resolve_sysimg_location, &ifo)) {
+        jl_error("fatal error: unexpected error while retrieving libpath");
+    }
+    strncpy(free_path, ifo.dli_fname, PATH_MAX);
+    path_size = strlen(ifo.dli_fname);
+#endif
+    if (path_size >= PATH_MAX) {
+        jl_error("fatal error: jl_options.julia_lib path too long");
+    }
+    jl_options.julia_lib = abspath(free_path);
+
+    // set julia_home
     if (!jl_options.julia_home) {
         jl_options.julia_home = getenv("JULIA_HOME");
         if (!jl_options.julia_home) {
-            jl_options.julia_home = dirname(free_path);
+            strncpy(free_path, jl_options.julia_lib, PATH_MAX);
+            free_path = dirname(dirname(free_path));
+            if (strlen(free_path)+4 >= PATH_MAX) {
+                jl_error("fatal error: jl_options.julia_home path too long");
+            }
+            strcat(free_path, PATHSEPSTRING "bin");
```

Comment:
I'm all for rethinking how `JULIA_HOME` is set, or even what it should mean (_e.g._, to me it would make a bit more sense to refer to one level up), but this is guessing code. It only gets invoked when someone _doesn't_ set their own `JULIA_HOME`. It works (and is only intended to work) in the usual case, with a standard installation.

Or do I somehow miss your point?